timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/database/Neo4jLuceneEntityFetcher.java

```diff
@@ -1,114 +1,71 @@
-  private final IndexManager indexManager;
-  private TinkerpopGraphManager graphManager;
+  private final IndexHandler indexHandler;
 
-  public Neo4jLuceneEntityFetcher(TinkerpopGraphManager graphManager) {
-    this.graphManager = graphManager;
+  public Neo4jLuceneEntityFetcher(TinkerpopGraphManager graphManager, IndexHandler indexHandler) {
     this.graphDatabase = graphManager.getGraphDatabase();
-    this.indexManager = graphDatabase.index();
-
+    this.indexHandler = indexHandler;
   }
 
   public GraphTraversal<Vertex, Vertex> getEntity(GraphTraversalSource source, UUID id, Integer rev,
                                                   String collectionName) {
     if (rev == null) {
-
       Optional<Vertex> foundVertex = getVertexByIndex(source, id, collectionName);
 
       if (foundVertex.isPresent()) {
         return source.V(foundVertex.get().id());
-      } else {
-        return super.getEntity(source, id, rev, collectionName);
       }
     }
     return super.getEntity(source, id, rev, collectionName);
   }
 
   private Optional<Vertex> getVertexByIndex(GraphTraversalSource source, UUID id, String collectionName) {
-    final Graph graph = graphManager.getGraph();
-
-    // Next comes some defensive programming (often introduces new problems in and of itself of course)...
-
-    // Open transaction to be able to access lucene indices
-    Transaction transaction = graph.tx();
-
-    if (!transaction.isOpen()) {
-      // Will throw: org.neo4j.graphdb.NotInTransactionException: null if transaction is not explicitly opened
-      transaction.open();
-    }
+    // Look up the vertex for this neo4j Node
+    GraphTraversal<Vertex, Vertex> vertexT = indexHandler.findById(id);
 
-    // Return when index does not exist
-    if (!indexManager.existsForNodes(collectionName)) {
-      transaction.close();
+    // Return if the neo4j Node ID matches no vertex (extreme edge case)
+    if (!vertexT.hasNext()) {
+      LOG.error(Logmarkers.databaseInvariant,
+        "Vertex with tim_id {} is found in index with id {}L but not in graph database", id);
       return Optional.empty();
     }
 
-    // Look up this uuid in the lucene index
-    Long vertexId = null;
-    IndexHits<Node> indexHits = indexManager.forNodes(collectionName).get("tim_id", id.toString());
-    if (indexHits.size() > 0) {
-      vertexId = indexHits.next().getId();
-    }
-
-    // Close the transaction before returning
-    transaction.close();
-
-    // Only continue when a neo4j Node was found
-    if (vertexId != null) {
-
-      // Look up the vertex for this neo4j Node
-      GraphTraversal<Vertex, Vertex> vertexT = source.V(vertexId);
-
-      // Return if the neo4j Node ID matches no vertex (extreme edge case)
-      if (!vertexT.hasNext()) {
-        LOG.error(Logmarkers.databaseInvariant,
-          "Vertex with tim_id {} is found in index with id {}L but not in graph database", id, vertexId);
+    // Get the latest version of the found Vertex
```

Comment:
Is this still needed? We make sure the index has the latest version anyway. And this code does not stop race conditions